File: desktop/src/data/notes.ts
// Typed wrappers over /v1/notes/* (internal/api/v1/notes.go, docs/api.md's
// "Notes (writable)" section) - the recorder's own second, isolated store
// for operator notes, feedback and (opt-in) follow-up threads. Reads go
// through agentGet (the record's own read path); every write goes through
// agentRequest (data/ai.ts), the one write surface the API has.

import { agentGet, ApiError } from "./tauri";
import { agentRequest } from "./ai";

/** internal/notes.Outcome's JSON values. */
export type NotesOutcome = "confirmed" | "false_positive" | "unresolved";

/** internal/notes.Annotation's JSON shape. */
export interface NotesAnnotation {
  incident_id: string;
  fingerprint: string;
  rule_id: string;
  root_cause_kind: string;
  root_cause_entity: string;
  opened_at_ns: number;
  outcome: NotesOutcome;
  cause_note?: string;
  resolution_note?: string;
  created_at_ms: number;
  updated_at_ms: number;
}

/** internal/notes.ThreadTurn's JSON shape. */
export interface NotesThreadTurn {
  at_ms: number;
  answer_id: string;
  question_redacted: string;
  summary_redacted: string;
}

/** internal/notes.Stats's JSON shape - GET /v1/notes/stats, what
 *  Diagnostics shows: counts and an on-disk size, never a path. */
export interface NotesStats {
  annotations: number;
  feedback: number;
  threads: number;
  bytes: number;
}

/** GET /v1/notes/incidents/{id}; null on a 404 (no note yet - not an
 *  error a caller needs to catch, matching agentGetRaw's own 304 idiom of
 *  a typed "nothing here" result instead of a thrown ApiError). */
export async function getNotesAnnotation(endpoint: string, incidentId: string): Promise<NotesAnnotation | null> {
  try {
    return await agentGet<NotesAnnotation>(endpoint, `/v1/notes/incidents/${encodeURIComponent(incidentId)}`);
  } catch (e) {
    if (e instanceof ApiError && e.status === 404) return null;
    throw e;
  }
}

export interface PutNotesAnnotationInput {
  ruleId: string;
  rootCauseKind: string;
  rootCauseEntity: string;
  openedAtNs: number;
  outcome: NotesOutcome;
  causeNote?: string;
  resolutionNote?: string;
}

/** PUT /v1/notes/incidents/{id} - creates or replaces the note, returning
 *  the stored row as confirmation (the API's own PUT response shape). */
export function putNotesAnnotation(
  endpoint: string,
  incidentId: string,
  input: PutNotesAnnotationInput,
): Promise<NotesAnnotation | null> {
  return agentRequest<NotesAnnotation>(endpoint, "PUT", `/v1/notes/incidents/${encodeURIComponent(incidentId)}`, {
    rule_id: input.ruleId,
    root_cause_kind: input.rootCauseKind,
    root_cause_entity: input.rootCauseEntity,
    opened_at_ns: input.openedAtNs,
    outcome: input.outcome,
    cause_note: input.causeNote,
    resolution_note: input.resolutionNote,
  });
}

export function deleteNotesAnnotation(endpoint: string, incidentId: string): Promise<null> {
  return agentRequest<null>(endpoint, "DELETE", `/v1/notes/incidents/${encodeURIComponent(incidentId)}`);
}

export interface GetNotesSimilarParams {
  ruleId: string;
  rootCauseKind: string;
  entity?: string;
  exclude?: string;
  limit?: number;
}

/** GET /v1/notes/similar - prior annotated incidents for the same rule and
 *  root cause, same rule+kind+entity first, then same rule+kind, newest
 *  first within each tier. Never null (the API's own guarantee). */
export function getNotesSimilar(endpoint: string, params: GetNotesSimilarParams): Promise<NotesAnnotation[]> {
  const q = new URLSearchParams({ rule_id: params.ruleId, root_cause_kind: params.rootCauseKind });
  if (params.entity) q.set("entity", params.entity);
  if (params.exclude) q.set("exclude", params.exclude);
  if (params.limit) q.set("limit", String(params.limit));
  return agentGet<NotesAnnotation[]>(endpoint, `/v1/notes/similar?${q.toString()}`);
}

export interface PostNotesFeedbackInput {
  answerId: string;
  incidentId: string;
  profile: string;
  modelId: string;
  helpful: boolean;
  ruleId?: string;
  rootCauseKind?: string;
  rootCauseEntity?: string;
}

export function postNotesFeedback(endpoint: string, input: PostNotesFeedbackInput): Promise<null> {
  return agentRequest<null>(endpoint, "POST", "/v1/notes/feedback", {
    answer_id: input.answerId,
    incident_id: input.incidentId,
    profile: input.profile,
    model_id: input.modelId,
    helpful: input.helpful,
    rule_id: input.ruleId,
    root_cause_kind: input.rootCauseKind,
    root_cause_entity: input.rootCauseEntity,
  });
}

export function getNotesThread(endpoint: string, incidentId: string): Promise<NotesThreadTurn[]> {
  return agentGet<NotesThreadTurn[]>(endpoint, `/v1/notes/threads/${encodeURIComponent(incidentId)}`);
}

export function postNotesThreadAppend(
  endpoint: string,
  incidentId: string,
  turn: { answerId: string; questionRedacted: string; summaryRedacted: string },
): Promise<null> {
  return agentRequest<null>(endpoint, "POST", `/v1/notes/threads/${encodeURIComponent(incidentId)}`, {
    answer_id: turn.answerId,
    question_redacted: turn.questionRedacted,
    summary_redacted: turn.summaryRedacted,
  });
}

export function getNotesSettings(endpoint: string): Promise<{ history_opt_in: boolean }> {
  return agentGet<{ history_opt_in: boolean }>(endpoint, "/v1/notes/settings");
}

export function putNotesSettings(endpoint: string, historyOptIn: boolean): Promise<null> {
  return agentRequest<null>(endpoint, "PUT", "/v1/notes/settings", { history_opt_in: historyOptIn });
}

/** GET /v1/notes/stats - see NotesStats. */
export function getNotesStats(endpoint: string): Promise<NotesStats> {
  return agentGet<NotesStats>(endpoint, "/v1/notes/stats");
}

export function forgetAllNotes(endpoint: string): Promise<null> {
  return agentRequest<null>(endpoint, "DELETE", "/v1/notes");
}

File: internal/notes/notes.go
// Package notes is the operator's own annotations on incidents - confirmed
// cause, false positive, unresolved, free-text notes, helpful/not-helpful
// feedback on a local-AI answer, and (opt-in only) follow-up question
// threads. This is deliberately NOT part of internal/store: the record
// (events.db) is read-only over the local API by construction
// (internal/api/v1/server_test.go's TestWriteMethodsAreNotRouted), and that
// guarantee has to keep meaning what it says. Notes are a second, isolated
// SQLite file next to the record (see DefaultPath), writable only through
// this package's own small API - never through anything that also touches
// events.db.
//
// See docs/product/adr/0008-operator-notes-beside-the-record.md for why this
// exists and what it deliberately does not do: a note is never treated as
// evidence, never merged in from an imported bundle, and never referenced by
// an incident's own stable identity (incident IDs are fresh ULIDs per
// firing, internal/correlate/engine.go's build() - see Fingerprint).
package notes

import "context"

// Outcome is what the operator concluded about one incident's real cause,
// independent of what either the deterministic engine or a local-AI
// hypothesis said.
type Outcome string

const (
	OutcomeConfirmed     Outcome = "confirmed"
	OutcomeFalsePositive Outcome = "false_positive"
	OutcomeUnresolved    Outcome = "unresolved"
)

// Valid reports whether o is one of the three defined outcomes - an empty
// or unrecognised value is a request error, not silently accepted.
func (o Outcome) Valid() bool {
	switch o {
	case OutcomeConfirmed, OutcomeFalsePositive, OutcomeUnresolved:
		return true
	}
	return false
}

// Field length caps. Generous for a real operator note, small enough that
// nobody can turn this into an unbounded local key-value store.
const (
	MaxNoteFieldLen = 2000
	MaxFeedback     = 500 // oldest dropped once exceeded
	MaxThreadTurns  = 20  // per incident
)

// Annotation is one incident's operator note. Fingerprint (see Fingerprint)
// is what actually identifies "this same conclusion" across a daemon
// restart or a re-fired incident with a new ULID; IncidentID is kept too,
// as the exact row a specific analysis session was looking at.
type Annotation struct {
	IncidentID      string  `json:"incident_id"`
	Fingerprint     string  `json:"fingerprint"`
	RuleID          string  `json:"rule_id"`
	RootCauseKind   string  `json:"root_cause_kind"`
	RootCauseEntity string  `json:"root_cause_entity"`
	OpenedAtNS      int64   `json:"opened_at_ns"`
	Outcome         Outcome `json:"outcome"`
	CauseNote       string  `json:"cause_note,omitempty"`
	ResolutionNote  string  `json:"resolution_note,omitempty"`
	CreatedAtMS     int64   `json:"created_at_ms"`
	UpdatedAtMS     int64   `json:"updated_at_ms"`
}

// Feedback is one helpful/not-helpful signal on a local-AI answer. Never
// sent anywhere; this package only ever writes it to the local notes store.
type Feedback struct {
	AnswerID    string `json:"answer_id"`
	IncidentID  string `json:"incident_id"`
	Fingerprint string `json:"fingerprint"`
	Profile     string `json:"profile"`
	ModelID     string `json:"model_id"`
	Helpful     bool   `json:"helpful"`
	AtMS        int64  `json:"at_ms"`
}

// ThreadTurn is one already-redacted follow-up question/answer pair, kept
// only when the operator has opted into on-device history (see
// Store.SetHistoryOptIn). The caller (internal/ai) is responsible for
// redaction before this ever reaches the store - this package stores
// exactly the strings it is given and never inspects their content.
type ThreadTurn struct {
	AtMS             int64  `json:"at_ms"`
	AnswerID         string `json:"answer_id"`
	QuestionRedacted string `json:"question_redacted"`
	SummaryRedacted  string `json:"summary_redacted"`
}

// Store is the write surface for operator notes. Every method is scoped to
// this package's own database; none can reach events.db, and nothing in
// this interface accepts a store.Store or any other record-facing type.
type Store interface {
	GetAnnotation(ctx context.Context, incidentID string) (*Annotation, error)
	PutAnnotation(ctx context.Context, a Annotation) error
	DeleteAnnotation(ctx context.Context, incidentID string) error

	// Similar returns prior annotated incidents matching the same rule and
	// root cause, ranked same-rule+kind+entity first, then same-rule+kind,
	// newest first within each tier, excluding excludeIncidentID. Never
	// returns more than limit rows.
	Similar(ctx context.Context, ruleID, rootCauseKind, rootCauseEntity, excludeIncidentID string, limit int) ([]Annotation, error)

	// GetAnnotations is the bulk form of GetAnnotation, for a bundle export
	// (internal/bundle) folding in every note for the incidents already in
	// its window - one query instead of one GetAnnotation call per
	// incident. Silently skips any id with no annotation (an incident an
	// operator never annotated is not an error); never returns more rows
	// than distinct ids were asked for. Empty in, empty (not nil-error)
	// out.
	GetAnnotations(ctx context.Context, incidentIDs []string) ([]Annotation, error)

	PutFeedback(ctx context.Context, f Feedback) error

	// HistoryOptIn reports the current on-device-history setting; threads
	// are stored only while it is true, and turning it off clears every
	// stored thread immediately (SetHistoryOptIn's own doc comment).
	HistoryOptIn(ctx context.Context) (bool, error)
	// SetHistoryOptIn changes the setting. Setting it to false deletes all
	// stored threads as part of the same call - "off" means forgotten, not
	// merely "stop adding more".
	SetHistoryOptIn(ctx context.Context, on bool) error
	GetThread(ctx context.Context, incidentID string) ([]ThreadTurn, error)
	// AppendThread refuses with ErrHistoryDisabled when HistoryOptIn is
	// false, rather than silently storing anonymous data nobody asked to
	// keep.
	AppendThread(ctx context.Context, incidentID string, turn ThreadTurn) error

	// ForgetAll deletes every annotation, feedback entry and thread. It
	// does not change the history-opt-in setting itself.
	ForgetAll(ctx context.Context) error

	// Stats reports counts and the on-disk size, for Diagnostics - never
	// the file path (which can carry a Windows username).
	Stats(ctx context.Context) (Stats, error)

	Close() error
}

// Stats is what Diagnostics shows about the notes store, deliberately
// nothing more specific than counts and a byte size.
type Stats struct {
	Annotations int   `json:"annotations"`
	Feedback    int   `json:"feedback"`
	Threads     int   `json:"threads"` // incidents with at least one stored turn
	Bytes       int64 `json:"bytes"`
}

// ErrHistoryDisabled is returned by AppendThread when HistoryOptIn is
// false.
var ErrHistoryDisabled = errHistoryDisabled{}

type errHistoryDisabled struct{}

func (errHistoryDisabled) Error() string {
	return "notes: history is not enabled; nothing was stored"
}
